src/Backend/ONE/ONEToolchain.ts

```diff
@@ -14,33 +14,106 @@
+    this.debianArch = DebianArch.amd64;
   }
 
   getToolchainTypes(): string[] {
     return this.toolchainTypes;
   }
 
-  getToolchains(_toolchainType: string, _start: number, _count: number): Toolchains {
-    throw Error('Invalid getToolchains call');
+  getToolchains(_toolchainType: string, _start?: number, _count?: number): Toolchains {
+    if (parseInt(execSync(`apt-cache show ${this.toolchainName} > /dev/null 2>&1; echo $?`)
+                     .toString()
+                     .trim()) !== 0) {
+      throw Error('the prerequisites are not met');
+    }
+
+    const availableToolchains = new Toolchains();
+
+    if (_toolchainType === 'latest') {
+      const versionList =
+          execSync(`apt-cache show ${this.toolchainName} | grep Version | awk '{print $2}'`)
+              .toString()
+              .trim()
+              .split('\n');
+
+      for (const version of versionList) {
+        const [major, minor, patch, option] = version.split(/[.~]+/);
+
+        const toolchainInfo = new ToolchainInfo(
+            this.toolchainName, 'Description: test',
+            new Version(parseInt(major), parseInt(minor), parseInt(patch), option));
+
+        const toolchain = new DebianToolchain(toolchainInfo, this.debianRepo, this.debianArch);
+        availableToolchains.push(toolchain);
+      }
+    }
+
+    return availableToolchains;
   }
 
   getInstalledToolchains(_toolchainType: string): Toolchains {
-    throw Error('Invalid getInstalledToolchains call');
+    const availableToolchains = new Toolchains();
+
+    if (_toolchainType === 'latest') {
+      const versionList =
+          execSync(`apt-cache show ${this.toolchainName} | grep Version | awk '{print $2}'`)
+              .toString()
+              .trim()
+              .split('\n');
+
+      for (const version of versionList) {
+        const [major, minor, patch, option] = version.split(/[.~]+/);
+
+        const toolchainInfo = new ToolchainInfo(
+            this.toolchainName, 'Description: test',
+            new Version(parseInt(major), parseInt(minor), parseInt(patch), option));
+
+        const toolchain = new DebianToolchain(toolchainInfo, this.debianRepo, this.debianArch);
+        availableToolchains.push(toolchain);
+      }
+    }
+
+    const installedToolchains = availableToolchains.filter((toolchain) => {
+      return parseInt(execSync(`dpkg-query --show ${toolchain.info.name} > /dev/null 2>&1; echo $?`)
+                          .toString()
+                          .trim()) === 0;
+    });
+    return installedToolchains;
   }
 
   prerequisitesForGetToolchains(): Command {
-    throw Error('Invalid prerequisitesForGetToolchains call');
+    const cmd = new Command('add-apt-repository');
+    cmd.push('ppa:psalmist-kim/ppa');
+    cmd.push('-y');
+    cmd.push('&&');
+    cmd.push('apt-get');
+    cmd.push('update');
+    cmd.setRoot();
+    return cmd;
```

Comment:
Please create a separate prerequisites script. I think `&&` is not working in ONE vscode.

Reply:
> Please create a separate prerequisites script. I think `&&` is not working in ONE vscode.

Is this because JavaScript and TypeScript have the same `&&` operator?

Reply:
No, it's depend on `ToolRunner` in ONE-vscode project. We can't use the `cp.spawn({shell: true})`, so we implemented a trick for running command with root privilege. This trick does not accept to use `&&` string, as I know. In my case, when I use this code, I got an error like this.

```
[9/26/2022, 1:38:30 PM][Env][info] Done building WorkFlow on Env: 
WorkJobs: [{"jobType":2,"name":"prerequisites","valid":true,"workDir":"/home/jyoung-linux","isCancelable":false,"tool":"add-apt-repository","toolArgs":["ppa:psalmist-kim/ppa","-y","&&","apt-get","update"],"root":true}]
[9/26/2022, 1:38:30 PM][Env][info] Showing password prompt
[9/26/2022, 1:38:34 PM][Env][info] Got password response
[9/26/2022, 1:38:34 PM][JobRunner][info] Run tool: add-apt-repository args: 
ToolArgs: ["ppa:psalmist-kim/ppa","-y","&&","apt-get","update"] cwd: /home/jyoung-linux root: true
[9/26/2022, 1:38:34 PM][ToolRunner][info] Running: prerequisites
[sudo] password for jyoung-linux: Error: need a single repository as argument
[9/26/2022, 1:38:34 PM][ToolRunner][info] child process exited with code 1
[9/26/2022, 1:38:34 PM][ToolRunner][info] Build Failed: 1

[9/26/2022, 1:38:34 PM][JobRunner][err] The job was failed.
[9/26/2022, 1:38:34 PM][ToolchainProvider][info] Installation was cancelled
```

Reply:
Thanks for the review. I've just checked. If I'm writing a separate script, where should I put the script? (I think you should be able to run the script even after building with the extension installation files. Is there a suitable location?)

Reply:
You can install `prerequisiteForGetToolchains.sh` script under the `script` directory.